### DISCUSSION TITLE
What is the process of building native extension?

### DISCUSSION THREAD

**Discussion by hurricup** [ruby/rubygems#5627] [Q&A] [answered]
I was thinking that it works like this:
1. `rubygems` unpacks `gem` archive into appropriate directory, e.g. `$GEM_PATH/gems/sqlite3-1.4.2`
2. puts gemspec into `$GEM_PATH/specifications/`
3. processes the `gemspec` and at some point runs `extconf.rb` specified in the `extensions` attribute of the `gemspec`
4. `extconf.rb` creates a `Makefile` inside the `$GEM_PATH/gems/sqlite3-1.4.2/ext` directory and invokes `make`
5. `make` builds the shared library and puts it into the `lib`. And that's it.

But what we have in real life:
1. we have a gem files in `$GEM_PATH/gems/sqlite3-1.4.2`
2. we have an object files, Makefile and shared library alongside the source files in `$GEM_PATH/gems/sqlite3-1.4.2/ext/`, but no building logs.
3. we have a shared library in the `lib` directory.
4. we have a separate `extensions` directory: `$GEM_PATH/extensions/x86_64-linux/3.1.0/sqlite3-1.4.2/` with a building logs (`gem.build_complete`, `gem_make.out`, `mkmf.log` ) and resulting shared library. 

So:
1. What is this `extensions` directory about?
2. Why do we need 3 copies of shared library? In `lib`, `ext`, and `extensions` directories?
3. Why `rubygems` adds both paths to the `$LOAD_PATH`: `$GEM_PATH/gems/sqlite3-1.4.2/lib` and `$GEM_PATH/extensions/x86_64-linux/3.1.0/sqlite3-1.4.2/` if we expecting to have shared library in `lib` directory?
4. For what reason shared library file may be missing in `lib` directory, but present in `ext`/`extensions` directories? (usually this happens on some system rubies)

**Comment by deivid-rodriguez** [comment, marked as answer]
Probably does not fully answer your question but gives some hint to investigate. I think there's some backwards compatibility stuff that makes the built extension live in several places:

https://github.com/rubygems/rubygems/blob/5d65b7f259bc06ec6749938f1bd6b72f8a4852a6/lib/rubygems/ext/ext_conf_builder.rb#L68-L75

(By the way that TODO should be moved to "RubyGems 4", because it was obviously not removed in RubyGems 3 😅.

**Reply by hurricup** [reply]
Is there some perspective on how it should work in future?

**Reply by hurricup** [reply]
Introduced in #749

**Reply by hurricup** [reply]
Seems the plan is to NOT install `so` files into the `lib`

**Comment by mohits** [comment]
I have been doing quite a bit of looking at how native gems are built in a series of blog posts: https://notepad.onghu.com/2023/learning-by-reversing-s1-native-gems/

I hope that this helps at least a bit?
